sw/host/tests/usbdev/usbdev_stream/usb_device.cc

```diff
@@ -0,0 +1,361 @@
+      // same device.
+      busSpec_ = busNumber_;
+      addrSpec_ = devAddress_;
+      found = true;
+    } else if (numTries-- > 0u) {
+      // Retry a number of times before reporting failure.
+      std::cout << '.' << std::flush;
+      sleep(1);
+    } else {
+      std::cerr << "Unable to locate USB device" << std::endl;
+      return false;
+    }
+  } while (!found);
+
+  // Report that we have at least found the device.
+  std::cout << "Device found (Bus " << (int)busNumber_ << " Device "
+            << (int)devAddress_ << ")" << std::endl;
+  if (verbose_) {
+    std::cout << " - Path: " << devPath_ << std::endl;
+  }
+
+  // We need to detach the kernel driver and claim the interface to have maximal
+  // control, eg. suspending device.
+  int rc = libusb_set_auto_detach_kernel_driver(devh_, 1u);
+  if (rc < 0) {
+    std::cerr << "Error detaching kernel driver: " << libusb_error_name(rc)
+              << std::endl;
+    return false;
+  }
+
+  // Read and check the currently active configuration; this should just be 1
+  // since our test software sets up only a single configuration.
+  int config;
+  rc = libusb_get_configuration(devh_, &config);
+  if (rc < 0) {
+    std::cerr << "Error getting configuration: " << libusb_error_name(rc)
+              << std::endl;
+  }
+
+  std::cout << "Configuration: " << config << std::endl;
+#else
+  std::cout << "Running without libusb" << std::endl;
+  devh_ = true;
+#endif
+  return true;
+}
+
+// Close the device, if open.
+bool USBDevice::Close() {
+#if STREAMTEST_LIBUSB
+  if (devh_) {
+    libusb_close(devh_);
+    devh_ = nullptr;
+  }
+#else
+  devh_ = false;
+#endif
+  return true;
+}
+
+bool USBDevice::Service() {
+#if STREAMTEST_LIBUSB
+  struct timeval tv = {0};
+  int rc = libusb_handle_events_timeout(ctx_, &tv);
+  if (rc < 0) {
+    return ErrorUSB("ERROR: Handling events", rc);
+  }
+#endif
+  return true;
+}
+
+bool USBDevice::ReadTestDesc() {
+  std::cout << "Reading Test Descriptor" << std::endl;
+
+  if (!Open()) {
+    return false;
+  }
+#if STREAMTEST_LIBUSB
+  // Send a Vendor-Specific command to read the test descriptor.
+  uint8_t testDesc[0x10u];
+  int rc =
+      libusb_control_transfer(devh_, 0xc2u, kVendorTestConfig, 0u, 0u, testDesc,
+                              sizeof(testDesc), kControlTransferTimeout);
+  if (rc < 0) {
+    std::cerr << "Error reading test descriptor: " << libusb_error_name(rc)
+              << std::endl;
+    return false;
+  }
+
+  if (verbose_) {
+    std::cout << "Test Descriptor:" << std::endl;
+    for (unsigned idx = 0u; idx < sizeof(testDesc); idx++) {
+      printf("%u: 0x%02x\n", idx, testDesc[idx]);
+    }
+  }
+
+  // Validate the received test descriptor.
+  const uint8_t test_sig_head[] = {0x7eu, 0x57u, 0xc0u, 0xf1u};
+  const uint8_t test_sig_tail[] = {0x1fu, 0x0cu, 0x75u, 0xe7u};
+  const uint8_t *dp = testDesc;
+  if (!memcmp(dp, test_sig_head, 4) && 0x10u == get_le16(&dp[4]) &&
```

Comment:
I wonder if having a `struct` would make this a bit easier to follow?

Reply:
I think it's best left as a simple byte stream, and that is how it's constructed on the device side (sw/device/lib/testing/usb_testutils_controlep.c). It's intended only to be a simple description of the test so that the host side can create appropriate streams, without running into potential endianness issues.